llmfoundry/data/finetuning/dataloader.py

```diff
@@ -199,28 +159,92 @@ def build_finetuning_dataloader(cfg: DictConfig,
                         +
                         f'of samples in your dataset to at least {minimum_dataset_size}.'
                     )
-
-        assert dataset is not None
-        dl = DataLoader(
-            dataset,
-            collate_fn=collate_fn,
-            batch_size=dataloader_batch_size,
+        sampler = dist.get_sampler(dataset,
             drop_last=cfg.drop_last,
-            sampler=dist.get_sampler(dataset,
-                                     drop_last=cfg.drop_last,
-                                     shuffle=cfg.dataset.shuffle),
-            num_workers=cfg.num_workers,
-            pin_memory=cfg.get('pin_memory', True),
-            prefetch_factor=cfg.get('prefetch_factor', 2),
-            persistent_workers=cfg.get('persistent_workers', True),
-            timeout=cfg.get('timeout', 0),
-        )
+            shuffle=cfg.dataset.shuffle)
+    else:
+        # Build streaming dataset
+        dataset = _build_streaming_dataset(cfg.dataset, tokenizer, device_batch_size)
+    
+    assert dataset is not None
```

Comment:
nit: Please avoid `assert` statement in the source code. It's okay to put it in the tests code. Instead, raise an Exception with the descriptive error message.

Reply:
I think we do have some asserts like this one for pyright. dataset is assigned for each conditional branch, but pyright is unable to infer that the type is not None